src/transformers/models/aria/modular_aria.py

```diff
@@ -0,0 +1,1357 @@
+
+        original_dtype = top_indices.dtype
+
+        tokens_per_expert = torch.histc(
+            top_indices.flatten().to(torch.float32),
+            bins=self.config.moe_num_experts,
+            min=0,
+            max=self.config.moe_num_experts - 1,
+        )
+
+        return scores, top_indices, tokens_per_expert.to(original_dtype)
+
+
+class AriaSharedExpertsMLP(LlamaMLP):
+    """
+    Shared Expert MLP for shared experts.
+
+    Unlike routed experts, shared experts process all tokens without routing.
+    This class reconfigures the intermediate size in comparison to the LlamaMLP.
+
+    Args:
+        config (AriaTextConfig): Configuration object for the Aria language model.
+    """
+
+    def __init__(self, config: AriaTextConfig):
+        nn.Module.__init__(self)
+        self.config = config
+        self.hidden_size = config.hidden_size
+        self.intermediate_size = config.moe_intermediate_size * config.moe_num_shared_experts
+        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=config.mlp_bias)
+        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=config.mlp_bias)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=config.mlp_bias)
+        self.act_fn = ACT2FN[config.hidden_act]
+
+
+class AriaGroupedExpertsGEMM(nn.Module):
+    """
+    Grouped GEMM (General Matrix Multiplication) module for efficient expert computation.
+    This module utilizes the grouped_gemm library (https://github.com/fanshiqing/grouped_gemm)
+    for optimized performance. If the grouped_gemm library is not installed, it gracefully
+    falls back to a sequential GEMM implementation, which may be slower but ensures
+    functionality.
+
+    Args:
+        in_features (int): Number of input features.
+        out_features (int): Number of output features.
+        groups (int): Number of expert groups.
+    """
+
+    def __init__(self, in_features, out_features, groups):
+        super().__init__()
+        self.in_features = in_features
+        self.out_features = out_features
+        self.groups = groups
+        self.weight = nn.Parameter(torch.empty(groups, in_features, out_features))
+
+    def forward(self, input, tokens_per_expert):
+        """
+        Perform grouped matrix multiplication.
+
+        Args:
+            input (torch.Tensor): Input tensor of shape (num_tokens, in_features).
+            tokens_per_expert (torch.Tensor): Number of tokens assigned to each expert.
+
+        Returns:
+            torch.Tensor: Output tensor of shape (num_tokens, out_features).
+        """
+        tokens_per_expert = tokens_per_expert.cpu()
+
+        # Ensure the CUDA device matches the input tensor's device.
+        # This mismatch can occur when using `transformers.AutoModel.from_pretrained`
+        # with `device_map="auto"` on a multi-GPU setup.
+        if torch.cuda.is_available():
+            torch.cuda.set_device(input.device)
```

Comment:
worth investigating and a fix in accelerate?

Reply:
let's:
1. use input.to(device) without a check
2. change no split module to account for this?